tests/phpunit/CRM/Civicase/Hook/BuildForm/TokenTreeTest.php

```diff
@@ -0,0 +1,357 @@
+    );
+    $this->assertEquals(
+      'Addressee ID',
+      $newTokenTree['Client']['children'][0]['children'][0]['text']
+    );
+    $this->assertEquals(
+      '{contact.custom_' . $this->contactCustomField['id'] . '}',
+      $newTokenTree['Client']['children'][1]['children'][0]['children'][0]['id']
+    );
+    $this->assertEquals(
+      $this->contactCustomField['name'],
+      $newTokenTree['Client']['children'][1]['children'][0]['children'][0]['text']
+    );
+  }
+
+  /**
+   * Verify case tokens.
+   *
+   * @param array $newTokenTree
+   *   List of tokens.
+   */
+  private function verifyCaseTokens(array $newTokenTree) {
+
+    $this->assertNotEmpty($newTokenTree['Case']);
+    $this->assertEquals(
+      '{case.id}',
+      $newTokenTree['Case']['children'][0]['children'][0]['id']
+    );
+    $this->assertEquals(
+      'Case Id',
+      $newTokenTree['Case']['children'][0]['children'][0]['text']
+    );
+    $this->assertEquals(
+      '{case.custom_' . $this->caseCustomField['id'] . '}',
+      $newTokenTree['Case']['children'][1]['children'][0]['children'][0]['id']
+    );
+    $this->assertEquals(
+      $this->caseCustomField['name'],
+      $newTokenTree['Case']['children'][1]['children'][0]['children'][0]['text']
+    );
+  }
+
+  /**
+   * Set contact related custom field.
+   */
+  private function setContactCustomFields() {
+    $contactGroupParams = [
+      'extends' => 'Contact',
+      'name' => 'test-contact' . uniqid(),
+      'title' => 'Test Contact' . uniqid(),
+    ];
+    $contactCustomGroup = civicrm_api3(
+      'CustomGroup',
+      'create',
+      $contactGroupParams
+    );
+    if (empty($contactCustomGroup['id'])) {
+      return;
+    }
+    $contactFieldParams = [
+      'custom_group_id' => $contactCustomGroup['id'],
+      'name' => 'test-case' . uniqid(),
+      'label' => 'Test Case' . uniqid(),
+      'data_type' => 'Boolean',
+      'default_value' => 1,
+      'html_type' => 'Radio',
+      'required' => 1,
+    ];
+    $contactCustomField = $this->createCustomField($contactFieldParams);
+    if (!empty($contactCustomField['id'])) {
+      $this->contactCustomField['id'] = $contactCustomField['id'];
+      $this->contactCustomField['name'] = $contactFieldParams['name'];
+    }
+  }
+
+  /**
+   * Set case related custom field.
+   */
+  private function setCaseCustomFields() {
+    $caseGroupParams = [
+      'extends' => 'Case',
+      'name' => 'test-case' . uniqid(),
+      'title' => 'Test Case' . uniqid(),
+    ];
+    $caseCustomGroup = civicrm_api3(
+      'CustomGroup',
+      'create',
+      $caseGroupParams
+    );
+    if (empty($caseCustomGroup['id'])) {
+      return;
+    }
+    $caseFieldParams = [
+      'custom_group_id' => $caseCustomGroup['id'],
+      'name' => 'test-case' . uniqid(),
+      'label' => 'Test Case' . uniqid(),
+      'data_type' => 'Boolean',
+      'default_value' => 1,
+      'html_type' => 'Radio',
+      'required' => 1,
+    ];
```

Comment:
The parameter passed to `createCustomField` here and in the `setContactCustomFields` function are very similar save the `custom_group_id`. Why don't we move these parameters to the `createCustomField` function as default params. I mentioned this too earlier.